### PR TITLE
[SPARK-16644][SQL] Aggregate should not propagate constraints containing aggregate expressions

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

aggregate expressions can only be executed inside `Aggregate`, if we propagate it up with constraints, the parent operator can not execute it and will fail at runtime.
## How was this patch tested?

new test in SQLQuerySuite

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @sameeragarwal @yhuai @liancheng

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 1]
```diff
@@ -2965,4 +2965,21 @@ class SQLQuerySuite extends QueryTest with SharedSQLContext {
       }
     }
   }
+
+  test("Aggregate should not propagate constraints containing aggregate expressions") {
```
Comment (rxin) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: put a jira number, and see if we can move this into constraint suite?

Comment (sameeragarwal) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Nice find. I agree, it'd be good to put this in `ConstraintPropagationSuite` and explicitly check for the inferred constraints.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/ConstraintPropagationSuite.scala [hunk 2]
```diff
@@ -79,13 +79,15 @@ class ConstraintPropagationSuite extends SparkFunSuite {
     assert(tr.analyze.constraints.isEmpty)
 
     val aliasedRelation = tr.where('c.attr > 10 && 'a.attr < 5)
-      .groupBy('a, 'c, 'b)('a, 'c.as("c1"), count('a).as("a3")).select('c1, 'a).analyze
+      .groupBy('a, 'c, 'b)('a, 'c.as("c1"), count('a).as("a3")).select('c1, 'a, 'a3).analyze
```
Comment (yhuai) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/ConstraintPropagationSuite.scala]: @sameeragarwal How's this one?

Comment (sameeragarwal) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/plans/ConstraintPropagationSuite.scala]: looks good

**Comment by sameeragarwal** [comment]
LGTM

**Comment by yhuai** [comment]
Thanks. I am merging this to master and branch 2.0.

**Pull request closed without merging**